packages/coinjoin/src/client/round/outputDecomposition.ts

```diff
@@ -167,20 +60,325 @@ export const getOutputAmounts = async (params: GetOutputAmountsParams) => {
+        round.id,
+        issuanceAmountCredentials,
+        issuanceVsizeCredentials,
+        zeroAmountCredentials,
+        zeroVsizeCredentials,
+        {
+            signal,
+            baseUrl: coordinatorUrl,
+            identity: getWeakRandomId(10),
+            delay: 0,
+            deadline: round.phaseDeadline,
+        },
+    );
+
+    // create credentials for output
+    const amountCredentialsOut = await middleware.getCredentials(
+        round.amountCredentialIssuerParameters,
+        issuanceData.realAmountCredentials,
+        issuanceAmountCredentials.credentialsResponseValidation,
+        { signal, baseUrl: middlewareUrl },
+    );
+    const vsizeCredentialsOut = await middleware.getCredentials(
+        round.vsizeCredentialIssuerParameters,
+        issuanceData.realVsizeCredentials,
+        issuanceVsizeCredentials.credentialsResponseValidation,
+        { signal, baseUrl: middlewareUrl },
+    );
+
+    // create zero credentials
+    const zeroAmountCredentialsOut = await middleware.getCredentials(
+        round.amountCredentialIssuerParameters,
+        issuanceData.zeroAmountCredentials,
+        zeroAmountCredentials.credentialsResponseValidation,
+        { signal, baseUrl: middlewareUrl },
+    );
+    const zeroVsizeCredentialsOut = await middleware.getCredentials(
+        round.vsizeCredentialIssuerParameters,
+        issuanceData.zeroVsizeCredentials,
+        zeroVsizeCredentials.credentialsResponseValidation,
+        { signal, baseUrl: middlewareUrl },
+    );
+
+    // return pairs of new credentials
+    // requested Credentials
+    const output = {
+        amountCredentials: [amountCredentialsOut[0], zeroAmountCredentialsOut[0]],
+        vsizeCredentials: [vsizeCredentialsOut[0], zeroVsizeCredentialsOut[0]],
+    };
+
+    // or change output should be returned to pool
+    // change Credentials
+    const change = {
+        amountCredentials: [amountCredentialsOut[1], zeroAmountCredentialsOut[1]],
+        vsizeCredentials: [vsizeCredentialsOut[1], zeroVsizeCredentialsOut[1]],
+    };
+
+    log(`Output amount credentials: ${output.amountCredentials.map(c => c.value).join(',')}`);
+    log(`Output vsize credentials: ${output.vsizeCredentials.map(c => c.value).join(',')}`);
+    log(`Change amount credentials: ${change.amountCredentials.map(c => c.value).join(',')}`);
+    log(`Change vsize credentials: ${change.vsizeCredentials.map(c => c.value).join(',')}`);
+
+    return {
+        output,
+        change,
+    };
+};
+
+const findCredentialsForTarget = (
+    target: number,
+    credentials: middleware.Credentials[],
+    maxValue: number,
+): [middleware.Credentials, middleware.Credentials] | undefined => {
+    // sort descending. higher possibility for match
+    const sorted = credentials.sort((a, b) => (a.value > b.value ? -1 : 1));
+
+    // find one Credential big enough to cover requested target
+    const bigCredential = sorted.find(cre => cre.value >= target);
```

Comment:
I think this `bigCredential` optimization sometimes fails to find the best pair as two smaller credentials could have bigger change without exceeding the limit. It should work without it equally well at worst.

Reply:
that's correct, it will find **first** matching value
i couldn't think about any reasonable fixtures to simulate a case where newly created change combined in next iteration was not enough create another input

any idea?

Reply:
I also cannot come up with any case where the smaller change breaks anything in the following iterations. So let's consider this sufficient.